### PR TITLE
Separate transaction protocol into a separate crate

### PR DESCRIPTION
**Pull request by altonen**
Introduce new crate called `sc-network-transactions` which contains the transaction protocol. Resolves the first item in #12133 

polkadot companion: https://github.com/paritytech/polkadot/pull/5973

### CODE REVIEW

File: client/finality-grandpa/src/lib.rs [hunk 1]
```diff
@@ -68,7 +68,9 @@ use sc_client_api::{
 	StorageProvider, TransactionFor,
 };
 use sc_consensus::BlockImport;
-use sc_network_common::protocol::ProtocolName;
+use sc_network_common::{
+    protocol::ProtocolName,
+};
```
Comment (dmitry-markin) [inline client/finality-grandpa/src/lib.rs:73]: `cargo +nightly fmt` should fix the formatting and make the CI happy

**Review by dmitry-markin — APPROVED** [review]
Looks good to me.

File: client/finality-grandpa/src/lib.rs (hunk 1 above)
Comment (altonen) [reply client/finality-grandpa/src/lib.rs:73]: Thanks, fixed

File: client/network/src/protocol/message.rs [hunk 2]
```diff
@@ -147,7 +144,7 @@ pub mod generic {
 		/// Block announce.
 		BlockAnnounce(BlockAnnounce<Header>),
 		/// Transactions.
-		Transactions(Transactions<Extrinsic>),
+		Transactions(sc_network_transactions::Transactions<Extrinsic>),
```
Comment (bkchr) [inline client/network/src/protocol/message.rs:147]: I checked the code and I don't think this is used anymore? If yes, could we please keep the variant here in the enum, but don't reference `sc_network_transactions`

File: client/network/src/service.rs [hunk 3]
```diff
@@ -1316,7 +1317,7 @@ where
 	/// that peer. Shared with the [`NetworkService`].
 	peers_notifications_sinks: Arc<Mutex<HashMap<(PeerId, ProtocolName), NotificationsSink>>>,
 	/// Controller for the handler of incoming and outgoing transactions.
-	tx_handler_controller: transactions::TransactionsHandlerController<H>,
+	tx_handler_controller: sc_network_transactions::TransactionsHandlerController<H>,
```
Comment (bkchr) [inline client/network/src/service.rs:1320]: This should be removed here as well. The idea is that the transaction protocol is not referenced anymore here anymore.

**Review by bkchr — COMMENTED** [review]
In general looking good, but I would like to have more separation :)

File: client/network/src/service.rs (hunk 3 above)
Comment (altonen) [reply client/network/src/service.rs:1320]: Should it be moved to be part of `Behaviour`?

**Comment by altonen** [comment]
@bkchr I removed the `sc-network-transactions` dependency from `sc-network` so now the networking code doesn't know anything about transactions (apart from some backwards-compatible things like `ServiceToWorkerMsg::PropagateTransactions` which we can can hopefully remove at some point).

There is one issue though which may be related to why CI is failing and which I haven't addressed yet: informing the transaction protocol about the status of major syncing. Previously it was done by the `sc-network` code but since it no longer can call the transaction protocol, we must find some other way of informing the transaction protocol about the state of major syncing.

The problem is that I can only come up with two ugly solutions: connect `sc-network-transactions` to `sc-network-sync` and when the syncing code decides that major syncing is done, it informs the transaction code about it through, e.g., a channel. Another idea would be to share `Arc<bool>` with the tasks. Do you or @dmitry-markin have something to suggest?

**Pull request closed without merging**